atoMEC/postprocess.py

```diff
@@ -1,78 +1,65 @@
+
+    .. math:: \mathrm{IPR}_i = \int \mathrm{d}{\mathbf{r}} |\Psi_i(\mathbf{r})|^4
+
+    It is typically used as a localization measure.
+
+    WARNING: the current definition in this version is not mathematically correct.
+    It does not include the proper contribution from the spherical harmonics
+    :math:`|Y_l^m(\theta,\phi)|^4`. This is omitted because it makes little difference
+    to the flavour of the results but complicates things. Currently, this function
+    does not correctly produce the expected limits (even if the spherical harmonic
+    contribution is correctly accounted for). Use at your own peril...
     """
-
-    def __init__(self, orbitals):
-
-        self.orbitals = orbitals
-        self._IPR_mat = np.zeros_like(orbitals.eigfuncs)
-        self._MIS = None
-
-    @property
-    def IPR_mat(self):
-        r"""ndarray: The inverse participation ratio matrix (of all orbitals)."""
-        if np.all(self._IPR_mat == 0.0):
-            self._IPR_mat = self.calc_IPR(self.orbitals.eigfuncs, self.orbitals._xgrid)
-        return self._IPR_mat
-
-    @property
-    def MIS(self):
-        r"""ndarray: The mean ionization state computed from the IPR."""
-        if self._MIS is None:
-            self._MIS = self.calc_MIS(self.orbitals, self.IPR_mat)
-        return self._MIS
-
-    @staticmethod
-    def calc_IPR(eigfuncs, xgrid):
-        r"""
-        Calculate the inverse participation ratio for all eigenfunctions.
-
-        Parameters
-        ----------
-        eigfuncs : ndarray
-            transformed radial KS orbitals :math:`P(x)=\exp(x/2)R(x)`
-        xgrid : ndarray
-            the logarithmic grid
-        """
-        # get the dimensions for the IPR matrix
-        spindims = np.shape(eigfuncs)[0]
-        lmax = np.shape(eigfuncs)[1]
-        nmax = np.shape(eigfuncs)[2]
-
-        IPR_mat = np.zeros((spindims, lmax, nmax))
-
-        # compute the fully delocalized limit (eigenfunctions constant)
-        V = 4.0 / 3.0 * pi * np.exp(xgrid[-1]) ** 3.0
-        psi_constant = V ** -2
-        Psi4_c_int = np.trapz(psi_constant * np.exp(3 * xgrid), xgrid)
-
-        # compute the IPR matrix
-        for i in range(spindims):
-            for l in range(lmax):
-                for n in range(nmax):
-                    Psi4 = eigfuncs[i, l, n, :] ** 4.0
-                    Psi4_int = np.trapz(Psi4 * np.exp(3 * xgrid), xgrid)
-
-                    # scale it by the fully delocalized limit
-                    IPR_mat[i, l, n] = 1 - Psi4_c_int / Psi4_int
-
-        return IPR_mat
+    # get the dimensions for the IPR matrix
+    spindims = np.shape(eigfuncs)[0]
+    lmax = np.shape(eigfuncs)[1]
+    nmax = np.shape(eigfuncs)[2]
+
+    IPR_mat = np.zeros((spindims, lmax, nmax))
+
+    # compute the IPR matrix
+    # FIXME: add spherical harmonic term
+    for i in range(spindims):
+        for l in range(lmax):
+            for n in range(nmax):
```

Comment:
These nested loops can cause performance bottlenecks and are potential candidates for vectorization (looping over numpy arrays should be avoided). For now, I've no idea if this is already performance critical or not....

Reply:
Mmm, luckily the product `nmax * lmax * spindims` is never that large (max few thousand) and the operations inside the loop are not expensive. (During tests on my laptop, things always ran basically instantaneously). I would say it's ok for now.